### PR TITLE
[Assist] Generalize embeddings system

### PR DESCRIPTION
**Pull request by xacrimon**

### CODE REVIEW

File: lib/ai/embeddingprocessor.go [hunk 1]
```diff
@@ -160,15 +156,15 @@ func NewEmbeddingProcessor(cfg *EmbeddingProcessorConfig) *EmbeddingProcessor {
...
-	node types.Server
-	data string
+// resourceStringPair is a helper struct that pairs a reosurce with a data string.
+type resourceStringPair struct {
+	resource types.Resource
+	data     string
 }
 
 // mapProcessFn is a helper function that maps a slice of nodeStringPair,
 // compute embeddings and return them as a slice of ai.Embedding.
```
Comment (zmb3) [inline lib/ai/embeddingprocessor.go:166]:
```suggestion
// compute embeddings and return them as a slice.
```

**Review by zmb3 — APPROVED** [review]

File: lib/ai/embeddingprocessor.go [hunk 2]
```diff
@@ -217,19 +213,35 @@ func (e *EmbeddingProcessor) process(ctx context.Context) {
...
+		return
+	}
+
+	idx := 0
+	resourceStream := stream.Func(func() (types.Resource, error) {
+		if idx == len(unifiedResources) {
+			return nil, io.EOF
+		}
+
+		resource := unifiedResources[idx].(types.Resource)
```
Comment (jakule) [inline lib/ai/embeddingprocessor.go:229]: I know this will never panic, but just in case if impossible happens, can you add the type check?

**Review by jakule — APPROVED** [review]

**Comment by jakule** [comment]
Also https://github.com/gravitational/teleport/actions/runs/5940756991/job/16110040068?pr=30835

**Comment by hugoShaka** [comment]
When implementing the embedder, we used backend streaming endpoints to appease the review gods and mitigate potential memory issues caused by requesting all Nodes at once (although this should not happen with our current Assist scope). To deal with streams we also implemented the zipper. This PR removes the backend streaming, but mimicks it by implementing a stream-like API after (while everything is still loaded in memory at the same time).

I'm totally OK with dropping streaming (although some people might disagree) but if we do, we should reconsider all the zipping logic. In our current state we end up both with the complexity of zipping streams and also the potential memory issues that streaming was supposed to avoid.

File: lib/ai/embeddingprocessor.go (hunk 2 above)
Comment (xacrimon) [reply lib/ai/embeddingprocessor.go:229]: sure

**Comment by xacrimon** [comment]
i'll simplify the zipping logic. the driving decision was that we seem to still keep the entire nodeset around in memory in auth in many places and that seems fine? it feels like we might want to consider something like an immutable (i.e snapshot-read) tree (of which go has a few impls) for a central cache as we'd get to avoid copying the set when returning from cache and it seems we still want to keep all nodes in memory for some things

**Pull request merged**